File: go/test/feature_flag_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestFeatureFlagEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.FeatureFlag(nil)
		if ent == nil {
			t.Fatal("expected non-nil FeatureFlagEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := feature_flagBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "feature_flag." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_FEATURE_FLAG_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		featureFlagRef01Ent := client.FeatureFlag(nil)
		featureFlagRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "feature_flag"}, setup.data), "feature_flag_ref01"))
		featureFlagRef01Data["project_id"] = setup.idmap["project01"]
		featureFlagRef01Data["unleash_id"] = setup.idmap["unleash01"]

		featureFlagRef01DataResult, err := featureFlagRef01Ent.Create(featureFlagRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		featureFlagRef01Data = core.ToMapAny(featureFlagRef01DataResult)
		if featureFlagRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// LOAD
		featureFlagRef01MatchDt0 := map[string]any{}
		featureFlagRef01DataDt0Loaded, err := featureFlagRef01Ent.Load(featureFlagRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if featureFlagRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

		// REMOVE
		featureFlagRef01MatchRm0 := map[string]any{
			"id": featureFlagRef01Data["id"],
		}
		_, err = featureFlagRef01Ent.Remove(featureFlagRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func feature_flagBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "feature_flag", "FeatureFlagTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read feature_flag test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse feature_flag test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"feature_flag01", "feature_flag02", "feature_flag03", "unleash01", "unleash02", "unleash03", "project01", "project02", "project03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_FEATURE_FLAG_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_FEATURE_FLAG_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_FEATURE_FLAG_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/FeatureFlagEntityTest.php
<?php
declare(strict_types=1);

// FeatureFlag entity test

require_once __DIR__ . '/../gitlab_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class FeatureFlagEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GitlabSDK::test(null, null);
        $ent = $testsdk->FeatureFlag(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = feature_flag_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "feature_flag." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_FEATURE_FLAG_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $feature_flag_ref01_ent = $client->FeatureFlag(null);
        $feature_flag_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.feature_flag"), "feature_flag_ref01"));
        $feature_flag_ref01_data["project_id"] = $setup["idmap"]["project01"];
        $feature_flag_ref01_data["unleash_id"] = $setup["idmap"]["unleash01"];

        $feature_flag_ref01_data_result = $feature_flag_ref01_ent->create($feature_flag_ref01_data, null);
        $feature_flag_ref01_data = Helpers::to_map($feature_flag_ref01_data_result);
        $this->assertNotNull($feature_flag_ref01_data);

        // LOAD
        $feature_flag_ref01_match_dt0 = [];
        $feature_flag_ref01_data_dt0_loaded = $feature_flag_ref01_ent->load($feature_flag_ref01_match_dt0, null);
        $this->assertNotNull($feature_flag_ref01_data_dt0_loaded);

        // REMOVE
        $feature_flag_ref01_match_rm0 = [
            "id" => $feature_flag_ref01_data["id"],
        ];
        $feature_flag_ref01_ent->remove($feature_flag_ref01_match_rm0, null);

    }
}

function feature_flag_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/feature_flag/FeatureFlagTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GitlabSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["feature_flag01", "feature_flag02", "feature_flag03", "unleash01", "unleash02", "unleash03", "project01", "project02", "project03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GITLAB_TEST_FEATURE_FLAG_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GITLAB_TEST_FEATURE_FLAG_ENTID" => $idmap,
        "GITLAB_TEST_LIVE" => "FALSE",
        "GITLAB_TEST_EXPLAIN" => "FALSE",
        "GITLAB_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GITLAB_TEST_FEATURE_FLAG_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["GITLAB_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["GITLAB_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new GitlabSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["GITLAB_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GITLAB_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
